patches/api/0416-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,758 @@
++
++        @NotNull
++        ArgumentType<UUID> uuid();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Entity>> entity();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Player>> player();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Collection<Entity>>> entities();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Collection<Player>>> players();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Collection<PlayerProfile>>> playerProfiles();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<BlockPosition>> blockPos();
++
++        @NotNull
++        ArgumentType<Component> component();
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..a4cdba3967813337e4c3a0a9ba8b1cd9e2769be8
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
+@@ -0,0 +1,86 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.mojang.brigadier.StringReader;
++import com.mojang.brigadier.arguments.ArgumentType;
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++
++/**
++ * An argument type that wraps around another argument.
++ * This argument is special in that the argument type that is wrapped will
++ * be sent to the client, not the wrapper itself.
++ *
++ * @param <T> type
++ * @param <B> base wrapping type
++ */
++@ApiStatus.Experimental
++public abstract class WrapperArgumentType<T, B> implements ArgumentType<T> {
++
++    @NotNull
++    private final ArgumentType<B> vanillaPrimitiveType;
++
++    public WrapperArgumentType(@NotNull ArgumentType<B> primitive) {
++        this.vanillaPrimitiveType = primitive;
++    }
++
++    /**
++     * Parses the argument using the wrapped argument type then converting
++     * it.
++     * @param reader string reader
++     * @return value
++     * @throws CommandSyntaxException if an error occurs while parsing
++     */
++    @NotNull
++    @Override
++    public T parse(@NotNull StringReader reader) throws CommandSyntaxException {
++        return this.convert(this.vanillaPrimitiveType.parse(reader));
++    }
++
++    /**
++     * Converts the type from the wrapped argument to this argument.
++     *
++     * @param baseType wrapped argument provided value
++     * @return converted value
++     * @throws CommandSyntaxException if an exception occurs while parsing
++     */
++    @NotNull
++    public abstract T convert(@NotNull B baseType) throws CommandSyntaxException;
++
++    /**
++     * Gets the wrapped argument.
++     *
++     * @return wrapped
++     */
++    @NotNull
++    public ArgumentType<B> getWrapped() {
```

Comment:
In the same vein as above, just getPrimitiveType() or smth instead of "wrapped". Similar to how the terminology in the PDC works.